plugins/function_replacement.c

```diff
@@ -0,0 +1,151 @@
+/*
+  This file is part of MAMBO, a low-overhead dynamic binary modification tool:
+      https://github.com/beehive-lab/mambo
+
+  Copyright 2018-2020 Guillermo Callaghan <guillermocallaghan at hotmail dot com>
+  Copyright 2018-2020 Unai Martinez-Corral <unai.martinezcorral at ehu dot eus>
+  Copyright 2018-2020 The University of Manchester
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+#ifdef PLUGINS_NEW
+
+#include <stdio.h>
+#include <assert.h>
+#include <locale.h>
+#include "../plugins.h"
+
+#define orig_name "rand"
+#define func_name rand_replacement
+#define func_argd void
+#define func_handler rand_handler
+
+extern void func_name(func_argd);
+
+int func_handler(mambo_context *ctx) {
```

Comment:
Is there any relevant implementation difference, or this is just a cleaner solution that didn't exist back then?

Reply:
This example uses the new API presented in the CC20 paper.

Reply:
There is actually a logic bug in the code: passing the actual fragment id to the inline hash table lookup routines could cause the dispatcher to (attempt) to patch the actual fragment exit. In this case, the now potentially incorrect exit code is unreachable, but depending on what checks and assertions are hit in the dispatcher, it could still cause issues.

But the main difference is that

* the public interfaces will remain stable for as long as possible, the private functions could get refactored or go away at any time
* this code will work on other RISC-like architectures we might support in the future

Reply:
Thanks a lot!